internals/src/macros.rs

```diff
@@ -37,3 +37,212 @@ macro_rules! impl_to_hex_from_lower_hex {
         }
     };
 }
+
+/// Creates a transparent wrapper around an inner type and soundly implements reference casts.
+///
+/// This macro takes care of several issues related to newtypes that need to allow casting their
+/// inner types to themselves:
+///
+/// * It makes sure to put repr(transparent) on the type
+/// * It optionally implements conversions from `&`, `&mut`, `Box`, `Rc`, `Arc`
+/// * It makes sure to put `#[inline]` on all of these conversions since they are trivial
+/// * It makes sure the reference cast is const
+/// * It makes sure the `Arc` conversion is conditioned on `target_has_atomic = "ptr"`
+///
+/// Usage: jut type the struct inside the macro as you would implementing it manually except leave
+/// `#[repr(transparent)]` out. Then add an impl block for the just-defined type containing a const
+/// function declaration that takes `&_` (use literal underscore) and returns `&Self`. Do not write
+/// the body, just semicolon.
+///
+/// Optionally add more such functions but use `&mut`, `Box` `Rc`, `Arc` instead of `&`. The
+/// functions must be in this order. The `alloc` types MUST NOT have import paths and don't need
```

Comment:
In a3c04532ce2a2bbbad26e58e5408e3173213be8b:

My read of the macro is that they don't actually need to be in order. Is this comment stale?

Reply:
Yes, originally I had it in order but then had to change it because of parsing ambiguity.